lib/ecto/query/builder/from.ex

```diff
@@ -58,18 +58,19 @@ defmodule Ecto.Query.Builder.From do
         schema when is_atom(schema) ->
           # Get the source at runtime so no unnecessary compile time
           # dependencies between modules are added
-          source = quote do: unquote(schema).__schema__(:source)
-          prefix = quote do: unquote(schema).__schema__(:prefix)
-          {1, query(prefix, source, schema)}
+          source = quote do: Ecto.Queryable.to_query(unquote(schema))
```

Comment:
The reason we had this code is because the query builder tries to generate as much as possible from the query at compilation time. By making this dynamic, we lose this flexibility.

Reply:
It should be fine to call it at compile-time, though. Should be basically the same as we do manually today.

Reply:
No because the optimizer only works if the query API we are invoking receives an `Ecto.Query` at *compilation time*. This is effectively hiding the query behind a runtime call, so it will all be lost.

Reply:
I'm saying we could call the protocol _at compile-time_. It shouldn't be different from doing what the implementation does manually.

Reply:
No, we cannot do that, because otherwise you can easily generate deadlocks. For example, imagine some query in MyApp.User using a join on Post and a query on MyApp.Post doing a join on User.

Reply:
Also the query returned by `__schema__(:query)` returns a value which we would need to escape and I am not sure we can guarantee all parameters in the query would actually be "escapable".

Plus it would add compile time dependencies between files, which we have generally avoided.

Reply:
@josevalim I definitely can get on board w/ not introducing compile time dependencies :D.